### PR TITLE
Divided user.js into seperate files for services and controllers.

### PR DESCRIPTION
**Pull request by amyvbenson**
Move hidUsers link function to a seperate controller and added unit tests.
Refactored so not making multiple calls to api to get users and to make userService simpler.

### CODE REVIEW

File: src/app/components/list/list.js [hunk 1]
```diff
@@ -122,8 +121,6 @@ listControllers.controller('ListCtrl', ['$scope', '$routeParams', '$location', '
       });
     };
     $scope.list.$promise.then(listCallback);
-    $scope.list.$httpPromise.then(listCallback);
```
Comment (guillaumev) [inline src/app/components/list/list.js:125]: Why was this removed ?

Comment (amyvbenson) [reply src/app/components/list/list.js:125]: It didn't seem necessary (I may well be wrong) and was making the users be requested from the api twice.

File: src/app/components/user/hidUsers.controller.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
+hidControllers.controller('hidUsersCtrl', ['$scope', '$rootScope', '$location', '$routeParams', 'gettextCatalog', 'alertService', 'hrinfoService', 'userService', 'User', 'List',
```
Comment (guillaumev) [inline src/app/components/user/hidUsers.controller.js]: I'm not a fan of naming a user related controller hidUsers and putting it into a hidUsers.controller.js file... Either we keep this as a directive, in which case it should go in a user.directives.js file, or we make it a controller, in which case it should go in user.controllers.js (at least for now, controllers can be split into multiple files later if needed)

Comment (amyvbenson) [reply src/app/components/user/hidUsers.controller.js]: I didn't add it to user.controllers.js as this is 615 lines long without it, which is far far too long already. 

It's the controller for the hidUsers directive so thought the name made sense.

File: src/app/components/list/list.js (hunk 1 above)
Comment (guillaumev) [reply src/app/components/list/list.js:125]: Ok so good catch on the fact that it makes the user request twice, however it is still needed and the issue needs to be fixed in another way. $promise.then is called only when the list has already been cached (in which case, with your removal, things would work). When the list hasn't already been cached however, $promise.then is NOT called (but $httpPromise.then is), which means that with your removal, when the list has not been cached already, listCallback would not be called.

File: src/app/components/list/list.js [hunk 3]
```diff
@@ -95,12 +95,11 @@ listControllers.controller('ListCtrl', ['$scope', '$routeParams', '$location', '
   $scope.isOwner = false;
   $scope.isFavorite = false;
 
-  $scope.$on('user-service-ready', function() { 
+  $scope.$on('user-service-ready', function() {
     $scope.list = List.get({'listId': $routeParams.list});
     var listCallback = function () {
       var req = [];
       req[$scope.list.type + 's.list'] = $scope.list._id;
-      userService.notify(req);
```
Comment (guillaumev) [inline src/app/components/list/list.js:103]: If this is removed, when the list is pulled from the API, how do you notify the user controller that it needs to update its list of users ?

File: src/app/components/list/list.js (hunk 1 above)
Comment (amyvbenson) [reply src/app/components/list/list.js:125]: ok thanks, I'll have a look at another way to deal with it.

File: src/app/components/user/hidUsers.controller.js (hunk 2 above)
Comment (guillaumev) [reply src/app/components/user/hidUsers.controller.js]: Ok good point, I don't know why I chose this name... Let's call it UsersCtrl and rename the existing UsersCtrl into UsersPageCtrl if that's ok with you.

File: src/app/components/list/list.js (hunk 3 above)
Comment (amyvbenson) [reply src/app/components/list/list.js:103]: The hidUsers controller calls getUsers, which populates the list. It didn't seem necessary to be going round in so many loops between the list and the user service and controller just to populate the list. But again there may be nuances I'm missing.

File: src/app/components/user/hidUsers.controller.js (hunk 2 above)
Comment (amyvbenson) [reply src/app/components/user/hidUsers.controller.js]: sure

File: src/app/components/list/list.js (hunk 1 above)
Comment (guillaumev) [reply src/app/components/list/list.js:125]: Ok I reread the documentation of angular-cached-resource and you're right, this can be removed, because $promise.then will be called no matter what (even if the list is not cached). Sorry about that.

File: src/app/components/user/hidUsers.controller.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      $scope.searchTerm = $routeParams.name;
+      var currentSortOrder = $scope.request.name;
+      var defaultRequest = {
+        limit: $scope.itemsPerPage,
+        offset: 0,
+        sort: 'name'
+      };
+
+      if ($routeParams.list) {
+        defaultRequest['lists.list'] = $routeParams.list;
```
Comment (guillaumev) [inline src/app/components/user/hidUsers.controller.js]: I think this will not work, because depending on the type of list a user is checked into, this can be operations.list, organizations.list etc... That's why I need the List to be resolved before calling getUsers, and that's why I need this notify (so that the list type is known and correct request can be made)

Comment (guillaumev) [reply src/app/components/user/hidUsers.controller.js]: That said, something to investigate might be to add all the lists under user.lists...
